db/db_bloom_filter_test.cc

```diff
@@ -1102,6 +1222,140 @@ TEST_P(DBFilterConstructionReserveMemoryTestWithParam, ReserveMemory) {
+        std::make_tuple(true, BloomFilterPolicy::Mode::kFastLocalBloom, false),
+        std::make_tuple(true, BloomFilterPolicy::Mode::kFastLocalBloom, true),
+        std::make_tuple(true, BloomFilterPolicy::Mode::kStandard128Ribbon,
+                        false),
+        std::make_tuple(true, BloomFilterPolicy::Mode::kStandard128Ribbon,
+                        true)));
+
+TEST_P(DBFilterConstructionCorruptionTestWithParam, DetectCorruption) {
+  Options options = CurrentOptions();
+  // We set write_buffer_size big enough so that flush won't be triggered before
+  // we manually trigger it for clean testing
+  options.write_buffer_size = 640 << 20;
+  BlockBasedTableOptions table_options = GetBlockBasedTableOptions();
+  options.table_factory.reset(NewBlockBasedTableFactory(table_options));
+  options.create_if_missing = true;
+  options.disable_auto_compactions = true;
+
+  DestroyAndReopen(options);
+  int num_key = static_cast<int>(GetNumKey());
+  Status s;
+
+  // Case 1: No corruption in filter construction
+  for (int i = 0; i < num_key; i++) {
+    ASSERT_OK(Put(Key(i), Key(i)));
+  }
+  s = Flush();
+  EXPECT_TRUE(s.ok());
+
+  // Case 2: Corruption of hash entries in filter construction
+  for (int i = 0; i < num_key; i++) {
+    ASSERT_OK(Put(Key(i), Key(i)));
+  }
+  SyncPoint::GetInstance()->SetCallBack(
+      "XXPH3FilterBitsBuilder::MaybeVerifyHashEntriesChecksum::PreVerification",
+      [&](void* arg) {
+        std::deque<uint64_t>* hash_entries_to_corrupt =
+            (std::deque<uint64_t>*)arg;
+        assert(!hash_entries_to_corrupt->empty());
+        *(hash_entries_to_corrupt->begin()) =
+            *(hash_entries_to_corrupt->begin()) ^ static_cast<uint64_t>(1);
+      });
+  SyncPoint::GetInstance()->EnableProcessing();
+
+  s = Flush();
+
+  if (table_options.detect_filter_construct_corruption) {
+    EXPECT_TRUE(s.IsCorruption());
+    EXPECT_TRUE(
+        s.ToString().find("Filter's hash entries checksum mismatched") !=
+        std::string::npos);
+  } else {
+    EXPECT_TRUE(s.ok());
+  }
+
+  SyncPoint::GetInstance()->DisableProcessing();
+  SyncPoint::GetInstance()->ClearCallBack(
+      "XXPH3FilterBitsBuilder::MaybeVerifyHashEntriesChecksum::"
+      "PreVerification");
+
+  // Case 3: Corruption of filter content in filter construction
+  DestroyAndReopen(options);
+
+  for (int i = 0; i < num_key; i++) {
+    ASSERT_OK(Put(Key(i), Key(i)));
+  }
+  std::string corrupted_filter_content;
+  Slice corrupted_filter;
+  SyncPoint::GetInstance()->SetCallBack(
+      "XXPH3FilterBitsBuilder::MaybePostVerify::PreVerification",
+      [&](void* arg) {
+        auto arg_pair =
+            (std::pair<std::unique_ptr<FilterBitsReader>*, std::string>*)arg;
+        std::unique_ptr<FilterBitsReader>* bits_reader_p = arg_pair->first;
+        std::string filter_content = arg_pair->second;
+
+        corrupted_filter_content = "@#$" + filter_content;
```

Comment:
This corruption re-arranges all the data, changing its size. I would prefer a more subtle corruption, such as clearing the first 8 bytes to zero. Not a big deal.